Python: feat: Add Agent Framework to A2A bridge support

Shubham-Kumar-2000 opened the pull request:
### Motivation and Context

This integration enables agent framework agents to communicate and execute as a A2A hosted agent server.

### Description

- Implement A2A event adapter for converting agent messages to A2A protocol
- Implement A2A executor for running agents in A2A environment
- Add comprehensive unit tests for event adapter, and executor
- Update agent framework core A2A module exports and type stubs
- Integrate thread management utilities for async execution
- Add getting started sample for A2A agent framework integration
- Update dependencies in uv.lock


### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [Contribution Guidelines](https://github.com/microsoft/agent-framework/blob/main/CONTRIBUTING.md)
- [x] All unit tests pass, and I have added new tests where possible
- [ ] **Is this a breaking change?** If yes, add "[BREAKING]" prefix to the title of the PR.

Shubham-Kumar-2000 committed:
feat: Add Agent Framework to A2A bridge support

- Implement A2A event adapter for converting agent messages to A2A protocol
- Add A2A execution context for managing agent execution state
- Implement A2A executor for running agents in A2A environment
- Add comprehensive unit tests for event adapter, execution context, and executor
- Update agent framework core A2A module exports and type stubs
- Integrate thread management utilities for async execution
- Add getting started sample for A2A agent framework integration
- Update dependencies in uv.lock

This integration enables agent framework agents to communicate and execute within the A2A (Agent to Agent) infrastructure.

Shubham-Kumar-2000 committed:
fix: Update references from agent_thread_storage to _agent_thread_storage in A2A executor tests

Shubham-Kumar-2000 committed:
Refactor A2A agent framework and improve code structure

- Reordered imports in various files for consistency and clarity.
- Updated `__all__` definitions to maintain a consistent order across modules.
- Simplified method signatures by removing unnecessary line breaks.
- Enhanced readability by adjusting formatting in several sections.
- Removed redundant comments and example scenarios in the execution context.
- Improved handling of agent messages in the event adapter.
- Added type hints for better clarity and type checking.
- Cleaned up test cases for better organization and readability.

Shubham-Kumar-2000 committed:
fix: Lint fix new line added

Shubham-Kumar-2000 committed:
test: Add unit tests for AgentThreadStorage and InMemoryAgentThreadStorage

Shubham-Kumar-2000 committed:
refactor: Update type hints to use new syntax for Union and List

Shubham-Kumar-2000 committed:
fix: Validate RequestContext for context_id and message before execution

Shubham-Kumar-2000 committed:
Refactor tests and remove A2aExecutionContext references

- Deleted the test file for A2aExecutionContext as it is no longer needed.
- Updated A2aExecutor tests to remove dependencies on A2aExecutionContext and adjusted method calls accordingly.
- Modified event adapter tests to use ChatMessage instead of AgentRunResponseUpdate.
- Removed A2aExecutionContext from imports in agent_framework.a2a module and updated type hints accordingly.

eavanvalkenburg commented on python/packages/a2a/agent_framework_a2a/_a2a_event_adapter.py:
```diff
@@ -0,0 +1,422 @@
+from agent_framework import ChatMessage, DataContent, Role, TextContent, UriContent
+
+
+class A2aEventAdapter(ABC):
```
Do we even need this base class? Since the base class also has this single method, it doesn't add anything.

eavanvalkenburg reviewed (commented):
I like the idea behind this, but there are some major issues with the implementation, especially the addition of the AgentThreadStore, please discuss that further with us.

Shubham-Kumar-2000 committed:
Refactor A2AExecutor tests and remove event adapter

- Updated test cases to use A2AExecutor instead of A2aExecutor for consistency.
- Removed mock_event_adapter fixture and related tests as A2aEventAdapter is deprecated.
- Consolidated event handling tests into TestA2AExecutorEventAdapter.
- Adjusted imports in various files to reflect the removal of deprecated components.
- Ensured all references to A2aExecutor are updated to A2AExecutor across the codebase.

Shubham-Kumar-2000 commented:
Hi @eavanvalkenburg thanks for the review.

I have further simplified the code for this feature and have addressed most of you review comments.

Regarding your major concern around `AgentThreadStore` I have started a discussion to finalize the approach.
https://github.com/microsoft/agent-framework/discussions/2463

Do let me know if any other changes are required.

Shubham-Kumar-2000 committed:
refactor: Remove AgentThreadStorage and InMemoryAgentThreadStorage classes from threads and tests

Shubham-Kumar-2000 committed:
feat: A2AExecutor to have its own override able save and get threads methods for persistent storage.

Shubham-Kumar-2000 committed:
fix: linter bugs

Shubham-Kumar-2000 committed:
Merge branch 'main' of https://github.com/Shubham-Kumar-2000/agent-framework into agent_framework_to_a2a

Shubham-Kumar-2000 committed:
removed unnecessary changes form core package

Shubham-Kumar-2000 committed:
new line added

Shubham-Kumar-2000 committed:
Merge branch 'main' of https://github.com/microsoft/agent-framework into agent_framework_to_a2a

Shubham-Kumar-2000 committed:
Refactor A2AExecutor tests and update imports

- Consolidated mock agent fixtures in test_a2a_executor.py to simplify agent mocking.
- Removed redundant tests related to thread storage and agent types, focusing on A2AExecutor's core functionality.
- Updated test assertions to reflect changes in message handling with new Message and Content classes.
- Enhanced integration tests to ensure compatibility with the new agent framework structure.
- Added A2AExecutor to the module exports in __init__.py and __init__.pyi for better accessibility.

Shubham-Kumar-2000 committed:
Update A2A documentation: enhance usage examples for A2AAgent and A2AExecutor

Shubham-Kumar-2000 committed:
Updated uv lock

Shubham-Kumar-2000 committed:
Fix metadata assertion in TestA2AExecutorHandleEvents and reorder load_dotenv call in agent_framework_to_a2a.py

Shubham-Kumar-2000 committed:
Update agent card configuration: add default input and output modes, and fix agent creation method

Shubham-Kumar-2000 committed:
Fix assertion for metadata in TestA2AExecutorHandleEvents

Shubham-Kumar-2000 committed:
Fix formatting issues in TestA2AExecutorExecute and TestA2AExecutorIntegration

Shubham-Kumar-2000 committed:
Enhance A2AExecutor documentation with examples and clarify agent execution process

Shubham-Kumar-2000 committed:
Merge branch 'main' into agent_framework_to_a2a

Shubham-Kumar-2000 committed:
Merge branch 'main' of https://github.com/microsoft/agent-framework into agent_framework_to_a2a

Shubham-Kumar-2000 committed:
Merge branch 'agent_framework_to_a2a' of https://github.com/Shubham-Kumar-2000/agent-framework into agent_framework_to_a2a

markwallace-microsoft commented:
<!-- Pytest Coverage Comment: python-test-coverage-report -->
<a href="https://github.com/microsoft/agent-framework/blob/6a5f29d4d4d82ff1631864c3f84b9202106224a7/README.md"><img alt="Python Test Coverage" src="https://img.shields.io/badge/Python Test Coverage-88%25-green.svg" /></a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>packages/a2a/agent_framework_a2a</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/6a5f29d4d4d82ff1631864c3f84b9202106224a7/packages/a2a/agent_framework_a2a/_a2a_executor.py">_a2a_executor.py</a></td><td>85</td><td>0</td><td>100%</td><td>&nbsp;</td></tr><tr><td><b>TOTAL</b></td><td><b>28410</b></td><td><b>3299</b></td><td><b>88%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 5683 | 30 :zzz: | 0 :x: | 0 :fire: | 1m 28s :stopwatch: |

Shubham-Kumar-2000 committed:
Merge branch 'main' of https://github.com/microsoft/agent-framework into agent_framework_to_a2a

Shubham-Kumar-2000 committed:
Revert uv lock to main

Shubham-Kumar-2000 committed:
Refactor A2AExecutor: Improve formatting and streamline constructor parameters

Shubham-Kumar-2000 committed:
Merge branch 'main' into agent_framework_to_a2a

Shubham-Kumar-2000 committed:
Merge branch 'main' into agent_framework_to_a2a

eavanvalkenburg commented:
@Shubham-Kumar-2000 sorry about the delays we've been busy with getting things to v1. Could you do another rebase to main and then I'll keep track and help you get this over the line!

Shubham-Kumar-2000 committed:
Merge branch 'main' into agent_framework_to_a2a

Shubham-Kumar-2000 commented:
Hi @eavanvalkenburg I have merged the master branch.

eavanvalkenburg commented on python/packages/a2a/agent_framework_a2a/_a2a_executor.py:
```diff
@@ -0,0 +1,195 @@
+        agent: The AI agent to execute.
+    """
+
+    def __init__(self, agent: Agent | WorkflowAgent):
```
we should use `SupportsAgentRun` protocol for this, both of these adhere to that and all other agents as well

eavanvalkenburg commented on python/packages/a2a/agent_framework_a2a/_a2a_executor.py:
```diff
@@ -0,0 +1,195 @@
+            user_message = Message(role="user", contents=[Content.from_text(text=query)])
+
+            # Run the agent with the message list
+            response = await self._agent.run(user_message, session=session)
```
nit: this would also be enough
```suggestion
            response = await self._agent.run(query, session=session)
```

eavanvalkenburg commented on python/packages/a2a/agent_framework_a2a/_a2a_executor.py:
```diff
@@ -0,0 +1,195 @@
+            user_message = Message(role="user", contents=[Content.from_text(text=query)])
+
+            # Run the agent with the message list
+            response = await self._agent.run(user_message, session=session)
```
when i look at our A2A connector, it seems it really only support streaming, should we use:

```suggestion
            response_stream = self._agent.run(user_message, stream=True, session=session)
```

and then use something like (with a updated handle_events function and a new `complete` closure that calls `updater.complete`)):
```python
stream = response_stream.with_transform_hook(self.handle_events).with_finalizer(complete)
await stream.get_final_response() # this ensures the whole stream is exhausted
```

Shubham-Kumar-2000 replied:
I have added support for both streaming and non-streaming modes

eavanvalkenburg commented on python/packages/a2a/agent_framework_a2a/_a2a_executor.py:
```diff
@@ -0,0 +1,195 @@
+                parts.append(Part(root=FilePart(file=FileWithBytes(bytes=base64_str, mime_type=content.media_type))))
+            elif content.type == "uri" and content.uri:
+                parts.append(Part(root=FilePart(file=FileWithUri(uri=content.uri, mime_type=content.media_type))))
+            # Silently skip unsupported content types
```
can we at least log those?

Shubham-Kumar-2000 replied:
added logs

eavanvalkenburg commented on python/packages/a2a/AGENTS.md:
```diff
@@ -4,20 +4,48 @@ Agent-to-Agent (A2A) protocol support for inter-agent communication.
 
-- **`A2AAgent`** - Agent wrapper that exposes an agent via the A2A protocol
+- **`A2AAgent`** - Client to connect to remote A2A-compliant agents.
+- **`A2AExecutor`** - Bridge to expose local agents via the A2A protocol.
```
```suggestion
- **`A2AExecutor`** - Bridge to expose Agent Framework agents via the A2A protocol.
```

eavanvalkenburg commented on python/samples/04-hosting/a2a/agent_framework_to_a2a.py:
```diff
@@ -0,0 +1,62 @@
+import uvicorn
```
missing copyright header

Shubham-Kumar-2000 replied:
Added copyright header

eavanvalkenburg commented on python/samples/04-hosting/a2a/agent_framework_to_a2a.py:
```diff
@@ -0,0 +1,62 @@
+
+    # --8<-- [start:AgentCard]
+    # This will be the public-facing agent card
+    public_agent_card = AgentCard(
```
could we add a helper function to `_a2a_executor` that generates this from a Agent object?

Shubham-Kumar-2000 replied:
We can but only fields available to us as part of agent being common to the `AgentCard` are `name` and `description` .  For all other fields, they will just become a parameter to the helper function as well.

We can still go ahead and add a simple helper function but not sure how much helpful it will be.

eavanvalkenburg commented on python/samples/04-hosting/a2a/agent_framework_to_a2a.py:
```diff
@@ -0,0 +1,62 @@
+    )
+    # --8<-- [end:AgentCard]
+
+    agent = OpenAIResponsesClient().as_agent(
```
we prefer the `Agent(client=...` pattern in the samples

eavanvalkenburg commented on python/samples/04-hosting/a2a/agent_framework_to_a2a.py:
```diff
@@ -0,0 +1,62 @@
+        defaultInputModes=['text'],
+        defaultOutputModes=['text'],
+        capabilities=AgentCapabilities(streaming=True),
+        skills=[skill],
```
I'm not too familair with the A2A spec, but it looks like the skill and agent cards have a lot of overlap? Why is that?

Shubham-Kumar-2000 replied:
I do agree that skill and agent card has a lot of over lap but those represent fundamentally 2 different things :

- Skill - Its like a capability / function that a agent advertises it can perform
- AgentCard - Its like a over all identity of a agent including all the skills.

For example a `Europe Travel Agent` can have skills like `Book flights` , `Book hotels`, etc.
In this case Agent card will be like `Europe Travel Agent` who specialises travel bookings in Europe.

As per as the Sample code let me update it with better example

eavanvalkenburg replied:
oh thanks for clarifying!

eavanvalkenburg commented on python/packages/a2a/agent_framework_a2a/_a2a_executor.py:
```diff
@@ -0,0 +1,195 @@
+            user_message = Message(role="user", contents=[Content.from_text(text=query)])
+
+            # Run the agent with the message list
+            response = await self._agent.run(user_message, session=session)
```
is there no way to get additional options passed from the caller?

Shubham-Kumar-2000 replied:
Added support to pass additional `run_kwargs`

eavanvalkenburg reviewed (commented):
Looks quite good overall, few small comments.

Shubham-Kumar-2000 committed:
Apply suggestions from code review

Co-authored-by: Eduard van Valkenburg <eavanvalkenburg@users.noreply.github.com>

Shubham-Kumar-2000 committed:
Refactor A2AExecutor to use SupportsAgentRun and enhance logging; update agent framework sample for flight and hotel booking capabilities

eavanvalkenburg commented on python/samples/04-hosting/a2a/agent_framework_to_a2a.py:
```diff
@@ -1,4 +1,9 @@
+from typing import Annotated
+
 import uvicorn
+from pydantic import Field
```
doesn't look like this is being used...

Shubham-Kumar-2000 replied:
removed

eavanvalkenburg committed:
Merge branch 'main' into agent_framework_to_a2a

Shubham-Kumar-2000 committed:
Enhance A2AExecutor with streaming support and custom run arguments; update tests for initialization and execution scenarios

Shubham-Kumar-2000 committed:
Merge branch 'agent_framework_to_a2a' of https://github.com/Shubham-Kumar-2000/agent-framework into agent_framework_to_a2a

Shubham-Kumar-2000 committed:
Enhance A2AExecutor event handling with streamed artifact tracking; update tests for new behavior

Shubham-Kumar-2000 commented:
Hi @eavanvalkenburg I have addressed all the review comments.

Please check once and let me know if any further changes are required from my end.

Shubham-Kumar-2000 committed:
Merge branch 'main' into agent_framework_to_a2a

eavanvalkenburg commented on python/packages/a2a/agent_framework_a2a/_a2a_executor.py:
```diff
@@ -0,0 +1,268 @@
+
+    async def _run_stream(self, query: Any, session: AgentSession, updater: TaskUpdater) -> None:
+        """Run the agent in streaming mode and publish updates to the task updater."""
+        response_stream = await self._agent.run(query, session=session, stream=True, **self._run_kwargs)
```
```suggestion
        response_stream = self._agent.run(query, session=session, stream=True, **self._run_kwargs)
```

Shubham-Kumar-2000 replied:
Fixed

eavanvalkenburg commented on python/packages/a2a/agent_framework_a2a/_a2a_executor.py:
```diff
@@ -0,0 +1,268 @@
+
+    async def _run(self, query: Any, session: AgentSession, updater: TaskUpdater) -> None:
+        """Run the agent in non-streaming mode and publish messages to the task updater."""
+        response = await self._agent.run(query, session=session, stream=False, **self._run_kwargs)
```
FYI: this is correct, it depends on the stream param

eavanvalkenburg commented on python/samples/04-hosting/a2a/agent_framework_to_a2a.py:
```diff
@@ -0,0 +1,75 @@
+    # --8<-- [end:AgentCard]
+
+    agent = Agent(
+        client=OpenAIChatCompletionClient(),
```
let's use at least `OpenAIChatClient` since that is the modern one, ideally we use FoundryChatClient in our samples, but you might not have that setup, so OpenAI is fine.

Shubham-Kumar-2000 replied:
Updated, I was actually using `gemini` and `OpenAIChatClient` doesn't seems to work with it that's why I was using chat completion client

Shubham-Kumar-2000 committed:
Refactor A2AExecutor to enforce type hints for stream and run_kwargs attributes

Shubham-Kumar-2000 committed:
Merge branch 'agent_framework_to_a2a' of https://github.com/Shubham-Kumar-2000/agent-framework into agent_framework_to_a2a

Shubham-Kumar-2000 committed:
Merge branch 'main' into agent_framework_to_a2a

Shubham-Kumar-2000 commented:
I have fixed the `mypy` issue as well

Shubham-Kumar-2000 committed:
Refactor A2AExecutor and tests: replace AsyncMock with MagicMock for response stream handling; clean up imports in agent_framework_to_a2a.py

Shubham-Kumar-2000 committed:
Merge branch 'agent_framework_to_a2a' of https://github.com/Shubham-Kumar-2000/agent-framework into agent_framework_to_a2a

Shubham-Kumar-2000 commented:
Hi @eavanvalkenburg I have addressed all new review comments.

Shubham-Kumar-2000 committed:
Merge branch 'main' into agent_framework_to_a2a

eavanvalkenburg commented:
@Shubham-Kumar-2000 one of the checks fail, please run the pre-commit jobs, it looks like a ruff issue

Shubham-Kumar-2000 committed:
refactor: streamline imports and improve code readability across multiple files

Shubham-Kumar-2000 committed:
Merge branch 'main' into agent_framework_to_a2a

Shubham-Kumar-2000 commented:
Hi @eavanvalkenburg I have fixed the lint issue and also merged with master.

Shubham-Kumar-2000 commented on python/packages/a2a/agent_framework_a2a/_a2a_executor.py:
```diff
@@ -0,0 +1,257 @@
+            return
+
+        parts: list[Part] = []
+        metadata = getattr(item, "additional_properties", None)
```
`metadata` is a nullable field A2A events

Shubham-Kumar-2000 commented on python/packages/a2a/agent_framework_a2a/_a2a_executor.py:
```diff
@@ -0,0 +1,257 @@
+            await updater.update_status(
+                state=TaskState.failed,
+                final=True,
+                message=updater.new_agent_message([Part(root=TextPart(text=str(e.args)))]),
```
I have observed is some python versions `str(e)` is aprearing as `object at 0x109f49280>`.

Shubham-Kumar-2000 replied:
Looks like this is no more the case.
I have fixed it

Shubham-Kumar-2000 committed:
feat: enhance A2AExecutor cancel method with context validation and fixed review comments

Shubham-Kumar-2000 commented on python/packages/a2a/agent_framework_a2a/_a2a_executor.py:
```diff
@@ -0,0 +1,257 @@
+        for content in contents:
+            if content.type == "text" and content.text:
+                parts.append(Part(root=TextPart(text=content.text)))
+            elif content.type == "data" and content.uri:
+                base64_str = content.uri
+                parts.append(Part(root=FilePart(file=FileWithBytes(bytes=base64_str, mime_type=content.media_type))))
+            elif content.type == "uri" and content.uri:
```
Fixed

Shubham-Kumar-2000 commented on python/packages/a2a/agent_framework_a2a/_a2a_executor.py:
```diff
@@ -0,0 +1,257 @@
+        Cancellation is primarily managed by the A2A protocol layer.
+        """
+        pass
+
```
Looks like in the latest version of A2A the overloaded function is responsible to call `updater.cancel()`

I have updated the implementation accordingly.

Shubham-Kumar-2000 committed:
feat: implement get_uri_data utility function for extracting base64 data from data URIs and update references

Shubham-Kumar-2000 committed:
fix: update import path for get_uri_data utility function in A2AExecutor and A2AAgent

Shubham-Kumar-2000 committed:
fix: correct error message handling in A2AExecutor and update test assertions

Shubham-Kumar-2000 committed:
Merge branch 'main' into agent_framework_to_a2a

Shubham-Kumar-2000 commented:
Hi @eavanvalkenburg 
This will require a re review from you as the previous one got stale due to copilot comments & fixes.
